### ISSUE TITLE
GPU delegate fails on Qwen2.5 models due to STABLEHLO_COMPOSITE: odml.softmax not supported by OpenCL/WebGPU delegate

### ISSUE THREAD

**Issue by bosejapan23-06** [google-ai-edge/litert-torch#1136]
## Summary
Exporting Qwen2.5-1.5B-Instruct via `litert-torch export_hf` produces a
.litertlm model that fails to fully delegate to GPU (OpenCL on Android/
Adreno, WebGPU on desktop) because softmax is packaged as a
STABLEHLO_COMPOSITE (`odml.softmax`) that the GPU delegate does not support.
Only ~16% of graph nodes delegate; the runtime is configured to hard-fail
when delegation isn't complete, so GPU backend initialization fails
entirely and the app falls back to CPU.

## Environment
- litert-torch-nightly: 0.10.0.dev20260725
- litert-lm: 0.14.0
- Export platform: Linux, x86_64 (also tested WebGPU/Vulkan locally, same failure)
- Target device: Google Pixel Tablet, Adreno GPU, OpenCL delegate
- Model: Qwen/Qwen2.5-1.5B-Instruct

## Steps to Reproduce
1. Export the model:
   litert-torch export_hf --model=Qwen/Qwen2.5-1.5B-Instruct \
     --output_dir=out --externalize_embedder \
     --prefill_lengths='[512]' --cache_length=4096 \
     --quantization_recipe=weight_only_wi8_afp32 \
     --enable_gpu_dynamic_prefill=True --enable_gpu_dynamic_cache=True

2. Run on GPU backend:
   litert-lm run out/model.litertlm --backend=gpu --prompt="Say Hello." --verbose

## Expected
Full (or near-full) graph delegation to GPU, matching the behavior of
Google's own published reference export
(litert-community/Qwen2.5-1.5B-Instruct), which delegates 1359/1359 and
1409/1409 nodes across its subgraphs with zero unsupported ops.

## Actual
ERROR: Following operations are not supported by GPU delegate:
STABLEHLO_COMPOSITE: odml.softmax
Replacing 233 out of 1434 node(s) with delegate (LITERT_WEBGPU) node,
yielding 2 partitions for subgraph 0.
ERROR: Hint fully delegated to single delegate is set, but the graph is
not fully delegated.
Failed to create engine: INTERNAL: ERROR: [.../llm_litert_compiled_model_executor.cc:1978]

On Android/Adreno with the OpenCL delegate, the identical failure occurs
with the same `odml.softmax` op at delegate_opencl.cc:330 during engine
initialization.

## Additional evidence: direct binary comparison

Extracted the tf_lite_prefill_decode section from both the failing model
and Google's own published reference model
(litert-community/Qwen2.5-1.5B-Instruct) as standalone .tflite files, and
searched for related strings:

$ strings ref_prefill_decode.tflite | grep -i "softmax\|composite\|stablehlo"
(zero matches — reference uses plain builtin SOFTMAX opcode)

$ strings ours_prefill_decode.tflite | grep -i "softmax\|composite\|stablehlo"
(54 distinct odml.softmax.impl_N symbols — one per attention layer per
exported signature)

This is a clean binary-level confirmation: the reference model contains
no STABLEHLO_COMPOSITE softmax wrapping at all, while every litert-torch
export we've produced bakes in odml.softmax composites.

## What we've already ruled out
- Recipe change (dynamic_wi8_afp32 vs weight_only_wi8_afp32): no effect on
  delegation count.
- --experimental_lightweight_conversion=True: no effect on delegation count
  (identical 233/1434 both with and without).
- Manually monkey-patching torch.nn.functional.softmax to decompose into
  primitive max/exp/sum/div ops before export, with attn_implementation=
  "eager" forced (to bypass fused SDPA): patch confirmed active (verified
  via call counter, fired 56 times during export matching 28 layers x 2
  signatures), but the exported graph still contains STABLEHLO_COMPOSITE:
  odml.softmax with identical 233/1434 delegation and identical
  odml.softmax.impl_N symbols in the binary. This strongly suggests the
  composite is re-introduced by the converter's own pattern-matching/
  fusion passes, not sourced from the model's PyTorch code.

## Question
Is there a supported way to either:
(a) prevent the converter from packaging softmax as a StableHLO composite
    for Qwen2/similar architectures, or
(b) have the GPU delegate accept a decomposed softmax lowering, or
(c) allow the runtime to proceed with partial delegation instead of
    hard-failing when "fully delegated" isn't achieved?

Happy to provide the full export log / verbose runtime log if useful.

**Comment by bosejapan23-06** [comment]
Could a maintainer please add the **bug** label to this?

**Comment by john-rocky** [comment]
Really thorough report — the binary diff against the reference export plus the call-counter monkeypatch test isolates this cleanly, and your conclusion is right: the composite is introduced by the exporter itself, not by the model's PyTorch code. I hit the same wall with a different model family (LFM2.5, litert-torch 0.9.2 + litert-lm 0.14.0 — same `odml.softmax` rejection, same fully-delegated hard-fail at engine creation, CPU unaffected), so sharing what worked for me on your option (a): an export-side workaround that gets full GPU delegation on the released runtime. Nothing here needs maintainer action — it's a user-side workaround, not a fix proposal.

Where the composite comes from: the exporter's attention module wraps the softmax it emits in `StableHLOCompositeBuilder(name="odml.softmax")` — `litert_torch/generative/export_hf/core/attention.py` (line ~93 on current main). That's why decomposing `torch.nn.functional.softmax` before export can't remove it: whatever softmax reaches that code path gets wrapped.

Workaround — neutralize the builder in that one module before invoking `export_hf`:

```python
from types import SimpleNamespace
from litert_torch.generative.export_hf.core import attention

class _PassthroughBuilder:
    def __init__(self, *args, **kwargs):
        pass
    def mark_inputs(self, *xs):
        return xs[0] if len(xs) == 1 else xs
    def mark_outputs(self, *xs):
        return xs[0] if len(xs) == 1 else xs

attention.composite = SimpleNamespace(StableHLOCompositeBuilder=_PassthroughBuilder)
```

Details, if useful:

- This strips only the composite *marker* (the grouping the released GPU delegates don't parse); the softmax computation itself stays in the graph unchanged, so the math is identical by construction.
- Rebinding the name only inside the `attention` module leaves other composites (e.g. `odml.rms_norm`) intact — those delegate fine.
- Verified on 0.9.2: an LFM2.5-1.2B int8 export goes from the identical engine-creation failure to fully delegated on WebGPU (M-series Mac, roughly 4.7k tok/s prefill / ~200 tok/s decode), with my output-quality gate matching the CPU path. The import path above is unchanged on current main, but I haven't run a 0.10 nightly — worth a quick sanity check on yours.
- One caveat worth stating: the marker presumably exists so that newer runtimes can pattern-match and fuse the whole softmax, so I'd treat this as a "released-runtime today" workaround rather than a better default. I keep it behind a `--keep-softmax-composite` escape hatch in my own export script for that reason (working example: [convert_lfm25.py](https://github.com/john-rocky/hf-to-litertlm/blob/main/lfm_work/convert_lfm25.py)).

(b) and (c) I'll leave to the maintainers — and if a supported switch for this already exists, I'd love to know too.

**Comment by shivvaam0001** [comment]
Hi @bosejapan23-06,

Thank you for bringing this to our attention.
Could you please provide the full export log or verbose runtime log? This will help us better understand and analyze the issue so we can provide a more accurate and helpful response.

Thanks!

**Comment by sirakiin** [comment]
The odml.softmax composite is due to an unsynchronized litert-converter release. 

Please use `pip install litert-converter==0.4.0.dev20260806` in the mean time to resolve the issue, and we are working on a new litert-converter release.

**Comment by john-rocky** [comment]
Confirmed — the dev converter build resolves it, thank you! With `litert-converter==0.4.0.dev20260806`, a composite-preserving export now runs fully on the GPU delegate on litert-lm 0.15.0.

Details, if useful:

- Setup: litert-torch 0.9.3 + litert-converter 0.4.0.dev20260806, exporting LiquidAI/LFM2.5-1.2B-Instruct (dynamic int8) with the odml.softmax composite kept.
- The converter now lowers the composite to the fused TFLite SOFTMAX builtin — the exported file carries zero odml.softmax markers, and the resulting graph is op-for-op identical to what the strip workaround above produced. So the workaround converges to exactly your fix, and can simply be retired once the release is out.
- GPU check on macOS (litert-lm 0.15.0): engine creation succeeds, prefill ~4.8k tok/s / decode ~269 tok/s on the 1.2B int8, and our 8-question quality spot-check passes 8/8 on both the CPU and GPU backends.
- One small install note for other readers: litert-torch 0.9.3 pins `litert-converter==0.3.*`, so the dev build needs a second step — `pip install litert-torch`, then `pip install --no-deps litert-converter==0.4.0.dev20260806`.

Nothing is needed from our side — once the new litert-converter release lands we'll switch our conversion script from stripping the composite to a simple converter-version check. Looking forward to the release, and thanks again for the quick follow-up.

**Comment by bosejapan23-06** [comment]
Thank you @sirakiin , @john-rocky. I will check and confirm.
Also please let me know if you have any solution for the below issue:
https://github.com/google-ai-edge/LiteRT-LM/issues/3083
